File: Arrays/best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.go
package main

func maxProfit(prices []int) int {
    maxProfit := 0
    minPrice := prices[0]
    
    for i := 1; i < len(prices); i++ {
        price := prices[i]
        minPrice = min(minPrice, price)
        maxProfit = max(maxProfit, price - minPrice)
    }
    
    return maxProfit
}

File: Arrays/best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.rs
use std::cmp;

pub struct Solution {}
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_profit = 0;
        let mut min_price = prices[0];

        for price in prices {
            min_price = cmp::min(min_price, price);
            max_profit = cmp::max(max_profit, price - min_price);
        }

        max_profit
    }
}

fn main() {
    let prices = vec![7, 1, 5, 3, 6, 4];
    let result = Solution::max_profit(prices);
    println!("{}", result);
}